plugin/documents.py

```diff
@@ -103,29 +112,34 @@ def _wrap_with_scope_style(self, content: str, scope: str, emphasize: bool = Fal
 
 
 class TextChangeListener(sublime_plugin.TextChangeListener):
-    def __init__(self, buffer: Any, listener: AbstractViewListener) -> None:
-        super().__init__(buffer)
-        self.listener = ref(listener)
 
-    # Created to have symmetry with "attach". Weird naming in internal API...
-    def detach(self) -> None:
-        self.remove()
+    ids_to_listeners = WeakValueDictionary()  # type: WeakValueDictionary[int, TextChangeListener]
+
+    @classmethod
+    def is_applicable(cls, buffer: sublime.Buffer) -> bool:
+        # FIXME: Cannot reliably use primary_view().file_name()
+        return buffer.primary_view().element() is None  # type: ignore
+
+    def __init__(self, buffer: sublime.Buffer) -> None:
+        super().__init__(buffer)
+        self.view_listeners = WeakSet()  # type: WeakSet[DocumentSyncListener]
+        self.ids_to_listeners[self.buffer.buffer_id] = self
 
     def on_text_changed_async(self, changes: Iterable[sublime.TextChange]) -> None:
-        listener = self.listener()
-        if listener:
+        for listener in list(self.view_listeners):
             listener.on_text_changed_async(changes)
 
     def on_reload_async(self) -> None:
-        listener = self.listener()
-        if listener:
+        for listener in list(self.view_listeners):
             listener.on_reload_async()
 
     def on_revert_async(self) -> None:
-        listener = self.listener()
-        if listener:
+        for listener in list(self.view_listeners):
             listener.on_revert_async()
 
+    def __repr__(self) -> str:
+        return str(self.buffer.buffer_id)
```

Comment:
If this is meant to help debugging then it would also worth adding that it's a TextChangeListener. Like
```
'TextChangeListener({}}'.format(self.buffer.buffer_id)
```